### PR TITLE
Add lotus API 

### PR DESCRIPTION
**Pull request by cryptoAtwill**
This PR adds the fundamental json rpc methods to interact with lotus node. At the same time, it contains a few simple fixes to the underlying `json_rpc` struct so that callers need to write fewer code. 
Changes includes:
- [x] Making `json_rpc::request` method handles both the `json_rpc` request obj and parsing response. Introduced deserialization of json directly to struct. Format code.
- [x] Added api commands such as wallet and state wait, read state, see `apis.rs`.

### CODE REVIEW

**Review by adlrocha — CHANGES_REQUESTED** [review]
Great job, @cryptoAtwill , thanks. I added a few suggestions around serialization, and it would be great if we can add some kind of basic test that we can use to test with a local Lotus node. Thanks!

File: src/main.rs
Comment (cryptoAtwill) [reply src/main.rs:3]: then it would be a lib

File: src/jsonrpc/api.rs
Comment (cryptoAtwill) [reply src/jsonrpc/api.rs:26]: I think `mpool_push_message` is better, since we are changing the struct to `LotusAPI`, then this is lotus specific, should be aligned with lotus api endpoint is easier to read. Also the function is declared as `async`, so I guess should be no need to specify `async`

File: src/jsonrpc/api.rs
Comment (cryptoAtwill) [reply src/jsonrpc/api.rs:76]: Yeah, I will do this once I have it working tested

File: src/jsonrpc/api.rs
Comment (cryptoAtwill) [reply src/jsonrpc/api.rs:50]: Initially I did this, but realized it is actually easier to just construct the `json`. `MpoolMessageRequest` should be internal to this method as well. The actual parameter to expose should be `MpoolPushMessage`, `MpoolMessageRequest` shall remain completely internal. Then I think construct the json is just convenient.

File: src/jsonrpc/api.rs
Comment (cryptoAtwill) [reply src/jsonrpc/api.rs:15]: I think const will have fewer abstraction? If converted to enum then need to convert to string anyways. I will put this into a mod maybe, then we can import from that mod as registered endpoint, the call pattern is very much like enum.

Comment (adlrocha) [reply src/jsonrpc/api.rs]: This sounds good!

**Comment by hmoniz** [comment]
If I understand correctly (the PR description could be more descriptive), this adds a more convenient and type-safe API to send requests to the Lotus JSON-RPC API, which is a laudable goal. You call this the `LotusApi`. Given this, I have the following comments:

- The `LotusApi` depends on the `JsonRpcClient`. The code organization should reflect this. First, have the `JsonRpcClient` be in its own independent module with proper visibility set in its members. Then, the `LotusApi` should be also a separate module that depends on the module of the `JsonRpcClient`. Right now, you have the `LotusApi` as a submodule of the `JsonRpcClient`, which is logically contradictory to the actual dependency relationship of these objects.

- Use a more descriptive name for the `LotusApi`. That struct ultimately is a client for sending JSON-RPC requests to a Lotus node. Maybe call it `LotusJsonRpcClient`, which more clearly signals the purpose of the struct.

- Define the `LotusApi` (or whatever we rename it to) as a trait. This will facilitate the testing of any function that depends on the `LotusApi` as it will easily allow the creation of a mock implementation and its injection.

- Please add good and descriptive documentation to the code. Particularly, to the public structs and functions that you expect other modules to use. For instance, the `LotusApi`, which will be a central component in the IPC agent because much of its logic depends on its use, should be thoroughly documented.

- When addressing these comments in your next iteration, please consider splitting the PR into smaller PRs, in which each one is self-contained and addresses one thing. This will substantially tighten the review cycle as your reviewers will need less context to review each PR. It will also make life easier for you as it will force you to make more structured changes to the code. Please, also add a really good description to each PR that clearly explains the what, how, and why of the change. If you keep your PRs small and logically self-contained to one "thing", this will be easy to do.

**Review by hmoniz — CHANGES_REQUESTED** [review]
If I understand correctly (the PR description could be more descriptive), this adds a more convenient and type-safe API to send requests to the Lotus JSON-RPC API, which is a laudable goal. You call this the `LotusApi`. Given this, I have the following comments:

- The `LotusApi` depends on the `JsonRpcClient`. The code organization should reflect this. First, have the `JsonRpcClient` be in its own independent module with proper visibility set in its members. Then, the `LotusApi` should be also a separate module that depends on the module of the `JsonRpcClient`. Right now, you have the `LotusApi` as a submodule of the `JsonRpcClient`, which is logically contradictory to the actual dependency relationship of these objects.

- Use a more descriptive name for the `LotusApi`. That struct ultimately is a client for sending JSON-RPC requests to a Lotus node. Maybe call it `LotusJsonRpcClient`, which more clearly signals the purpose of the struct.

- Define the `LotusApi` (or whatever we rename it to) as a trait. This will facilitate the testing of any function that depends on the `LotusApi` as it will easily allow the creation of a mock implementation and its injection.

- Please add good and descriptive documentation to the code. Particularly, to the public structs and functions that you expect other modules to use. For instance, the `LotusApi`, which will be a central component in the IPC agent because much of its logic depends on its use, should be thoroughly documented.

- When addressing these comments in your next iteration, please consider splitting the PR into smaller PRs, in which each one is self-contained and addresses one thing. This will substantially tighten the review cycle as your reviewers will need less context to review each PR. It will also make life easier for you as it will force you to make more structured changes to the code. Please, also add a really good description to each PR that clearly explains the what, how, and why of the change. If you keep your PRs small and logically self-contained to one "thing", this will be easy to do.

**Comment by cryptoAtwill** [comment]
@hmoniz I'll split into another module and add the interface. The function names and struct names are taken directly from lotus documentation and links are attached, I don't know what documentation can be added.

**Comment by hmoniz** [comment]
> @hmoniz I'll split into another module and add the interface. The function names and struct names are taken directly from lotus documentation and links are attached, I don't know what documentation can be added.

Thanks. In line with the best practices that we've agreed upon as a team, I think this PR is a good opportunity for you to get into the habit of creating small PRs. I suggest that you use [graphite](https://graphite.dev/docs/introduction) for this. It also has the functionality to help you split an existing PR like this one. I suggest splitting this into the following smaller PRs: 

- First PR with only the changes required in the JsonRpcClient
- Second PR with the `LotusJsonRpcClient` trait and a bare-bones implementation. For example, the implementation can have just one method implemented as an example for the reviewer.
- Third PR that adds the implementation of the remaining methods.

I'll hold off on the review until you make this split. Thanks!

File: src/lotus/client.rs
Comment (hmoniz) [inline src/lotus/client.rs:48]: I don't think that the nonce can be an optional parameter. Did you test this?

File: src/lotus/client.rs
Comment (hmoniz) [inline src/lotus/client.rs:53]: Please rename this with a descriptive name. Someone who's reading this shouldn't have to read the actual implementation to understand what the function does.

File: src/lotus/client.rs
Comment (hmoniz) [inline src/lotus/client.rs:63]: These are the `params` to the JSON_RPC request. Please rename appropriately, here and elsewhere through this file.

File: src/lotus/client.rs
Comment (hmoniz) [inline src/lotus/client.rs:128]: Please give this variable a more descriptive name.

File: src/lotus/types.rs
Comment (hmoniz) [inline src/lotus/types.rs:87]: This naming is very confusing. This is actually the meat of the response to `MpoolPushMessage` but this makes it sound like it's something else. I strongly suggest you namespace the types declared in this file so it can be clearly understood what structs are requests, which are responses, etc.

File: src/lotus/types.rs
Comment (hmoniz) [inline src/lotus/types.rs:1]: Please rename this file to something less generic. Using a `types.rs` file is considered an antipattern because it quickly degenerates into a dumping ground for random declarations without much concern for proper structure.

This file seems to be mostly dedicated to declaring message types used for the JSON-RPC methods. Please let it reflect that.

File: src/lotus/types.rs
Comment (hmoniz) [inline src/lotus/types.rs:181]: These functions shouldn't be conflated into the same file as the message declarations. Please move them to a more appropriate place.

File: src/lotus/types.rs
Comment (hmoniz) [inline src/lotus/types.rs:196]: Is this really necessary or just something you used to debug the code?

File: src/lotus/types.rs
Comment (hmoniz) [inline src/lotus/types.rs:9]: All these structs need proper doc comments.

**Review by hmoniz — CHANGES_REQUESTED** [review]

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:81]: This is debug assert, it wont trigger in the optimized build.

File: src/bin/sample.rs
Comment (cryptoAtwill) [reply src/bin/sample.rs:1]: this is example bin, why exclude this?

File: src/lotus/mod.rs
Comment (cryptoAtwill) [reply src/lotus/mod.rs:19]: There is no need to change name. I assume even if it's http/ws, it's still the same interface.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:29]: No need, the Inner should be used to reflect the type. You can implement `LotusWSClient` with the same struct.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:30]: If not typed, you mean pass in an interface or impl?

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:48]: yeah, it is optional

File: src/lotus/types.rs
Comment (cryptoAtwill) [reply src/lotus/types.rs:181]: what?

File: src/lotus/types.rs
Comment (cryptoAtwill) [reply src/lotus/types.rs:196]: yeah, to ensure I can get the same result as specified in the lotus docs.

File: src/jsonrpc/mod.rs
Comment (hmoniz) [reply src/jsonrpc/mod.rs:81]: If it it's included in a release build, then it means that we don't perform this error checking at all. Is that what you intend?

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:45]: `from` is not required even in the go cli. It's using default if not provided.

File: src/jsonrpc/mod.rs
Comment (hmoniz) [reply src/jsonrpc/mod.rs:32]: You're making this change. You need to assume ownership that the code it affects remains cohesive. It's totally understandable that you don't want to make this change as part of this PR, but you should then at least add a TODO comment and make sure that you or someone else follow up on it. It might require opening an issue in addition to the TODO.

File: src/bin/sample.rs
Comment (hmoniz) [reply src/bin/sample.rs:1]: Because it's not something we need in the production code and while it's checked in it's something that we're forced to maintain.

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:32]: I will add the TODO

File: src/bin/sample.rs
Comment (cryptoAtwill) [reply src/bin/sample.rs:1]: We can remove it later? Then how do you expect other to start in the first place?

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:81]: Only check in debug build.

Comment (hmoniz) [reply src/jsonrpc/mod.rs:82]: Can you explain why you would want this check in the debug build but not in the release?

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:30]: The only dependency of `LotusClient` is a `LotusApi` field. There is no need to parameterize its type.

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:29]: Regarding the `Inner`, please see also my other comment where I mention that it should not be needed. The current `JsonRpcClient` uses both HTTP and WS. You're not going to need, at least at this point, separate implementations for the different transport modes. But, even if you did, they would be behind the same trait.

For the naming, choose whatever you prefer as long as the name clearly signals that it is an implementation of the `LotusApi` trait (which should itself be renamed).

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:45]: A JSON-RPC method (e.g., MpoolPushMessage) stands on its own. It's simply a certain JSON-encoded HTTP/WS request sent to an endpoint. These methods (e.g., `mpool_push_message`) are a. way of sending such requests. If I call `mpool_push_message`, I expect a MpoolPushMessage JSON-RPC request to be sent over and only that request. With this implementation, we might be sending a WalletDefaultAddress JSON-RPC request in addition to the intended MpoolPushMessage request. If the user needs to call WalletDefaultAddress to obtain an address to be used as input to MpoolPushMessage, then it should call `wallet_deault()` explicitly and outside of `mpool_push_message()`.

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:82]: This is a request and reply, they are supposed to be the same. Actually, I even think we should tighten this to just panic. If the entries dont match, it's some series error.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:45]: so you are suggesting making it into two methods? One calling another? Something like `mpool_push_message_with_from` and `mpool_push_message_without_from` and `mpool_push_message_without_from` calls into `mpool_push_message_with_from`?

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:30]: If you use `JsonRpcClientImpl`, then what's the point of having that interface? I don't understand this.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:29]: `But, even if you did, they would be behind the same trait.` You mean they would or wouldn't? Currently the design is such that they will be sharing the same trait.

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:30]: Not `JsonRpcClientImpl`, but the `JsonRpcClient` trait. You depend on the interface, not the implementation.

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:45]: No, I'm asking you to simplify this method by not having `from` be optional.

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:48]: Thanks!

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:82]: Under what circumstances will this happen? Most likely that node you are requesting is having some trouble? This is definitely not the correct `json_rpc` behavior. Then shouldnt it just crash the program? Why do you think this is `recoverable`?

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:31]: You cannot do that in rust or you need to use `dyn` which is far more complex...

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:46]: Why? You mean force caller to get the default address every time?

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:34]: I think this is just a preference. I think Res should be fine.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:30]: I think we are roughly on the same page, but the thing is Rust has some quirks of its own and the above wont compile, also it has some `Send` + `Sync` complaints. To make it work and then will result in the current implementation.

Comment (hmoniz) [reply src/lotus/client.rs:30]: The above was just a sketch. Please look at this commit, where you can see it working:
https://github.com/consensus-shipyard/ipc-agent/commit/eeb8d7e5537906a3fb43d86b19609c5f4ed3a956

File: src/lotus/mod.rs
Comment (hmoniz) [reply src/lotus/mod.rs:19]: I'm sorry, but I'm going to have to insist that you rename this. It might not sound like a big deal, but this sort of thing adds up to the cognitive load required to understand the codebase. For instance, Lotus is riddled with types named in such a way (e.g., NetApi, FullNodeApi, WalletApi, etc.), making much of the code really difficult to read because of the lack of specificity in the naming. This trait is an interface for a JSON-RPC client that sends requests to a Lotus node. Please call it what it is, a `LotusJsonRpcClient`. By calling it `LotusApi`, with don't know if it's a client, a server, or something entirely different. We really don't know what to expect when seeing an object of such a type in the code since the name is just too generic.

File: src/lotus/client.rs
Comment (hmoniz) [reply src/lotus/client.rs:46]: The caller explicitly sets the `from` field in `MpoolPushMessage`. They can use whatever value they want for that field. Even if they want to use the default address, they don't have to get it every time. They can get it once and save it. In any case, the main reason I'm asking for this change is simple: a call to a JSON-RPC method (i.e., MpoolPushMessage) should not be calling other JSON-RPC methods under the hood (i.e., WalletDefault). Each method in this interface should be calling only the corresponding JSON-RPC method, nothing else.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:30]: So in the end it is just renaming?

Btw, I generally prefer putting less type restrictions on the generics. 

Also I can change this to `LotusJsonRPCClient`, but I just explain here why I didnt. I was thinking in the future it is possible to have maybe HTTP or TCP or other client. In this case, we could have:
```rust
pub struct LotusClient<T> {
    client: T,
}

impl <T: JsonRpcClient> LotusClient<T> {
    ...
}

impl <T: JsonHttpClient> LotusClient<T> {
    ...
}
```
So this allows multiple implementations and they can share the same api.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs]: I have already updated.

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:31]: I replied in that comment

File: src/lotus/client.rs
Comment (cryptoAtwill) [reply src/lotus/client.rs:46]: Sorry, why cannot call other RPC methods? The golang client does that?

Comment (cryptoAtwill) [reply src/lotus/client.rs:46]: What I'm trying to understand is this: `a call to a JSON-RPC method (i.e., MpoolPushMessage) should not be calling other JSON-RPC methods under the hood (i.e., WalletDefault). Each method in this interface should be calling only the corresponding JSON-RPC method, nothing else.` seems to be a very strict criteria? What if the method calls for calling multiple JSON-RPC?

File: src/jsonrpc/mod.rs
Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs:82]: Ok ok, I'll make the change. But I still think we should be very strict on this.

File: src/bin/sample.rs
Comment (cryptoAtwill) [reply src/bin/sample.rs]: I renamed the folder to examples, maybe that's more appropriate.

File: src/lotus/response.rs
Comment (adlrocha) [inline src/lotus/response.rs:38]: As mentioned sync, we should have a look at how forest does this. I think there may already be convenient methods to `serde` `cid` and other IPLD data types: https://github.com/ChainSafe/forest/blob/main/ipld/src/json.rs

File: src/bin/sample.rs
Comment (adlrocha) [reply src/bin/sample.rs]: You can add test documentation as comments instead of checking-in unnecessary files just to illustrate their use. See https://doc.rust-lang.org/stable/rustdoc/write-documentation/documentation-tests.html. For a library like this I would avoid having example execution files.

File: src/jsonrpc/mod.rs
Comment (jsoares) [reply src/jsonrpc/mod.rs]: Postel's law is your friend: 

https://en.wikipedia.org/wiki/Robustness_principle

Comment (cryptoAtwill) [reply src/jsonrpc/mod.rs]: already updated

**Comment by cryptoAtwill** [comment]
@hmoniz @adlrocha all comments updated, please take a look. Should be good to go.

File: src/lotus/client.rs
Comment (hmoniz) [inline src/lotus/client.rs:30]: `inner` is still a not very descriptive name. I suggest using `json_rpc_client` or just `client`.

File: src/lotus/message.rs
Comment (hmoniz) [inline src/lotus/message.rs:92]: If this is inside `MpoolPushMessageResponse`, then I suggest you name it `MpoolPushMessageResponseInner`.

File: src/lotus/mod.rs
Comment (hmoniz) [inline src/lotus/mod.rs:20]: Please drop "Api" from the name as it adds confusion to the intention. I suggest you rename it to `LotusJsonRpcClient` as suggested before or just `LotusClient`, though the former is more descriptive.

**Review by hmoniz — CHANGES_REQUESTED** [review]

**Comment by cryptoAtwill** [comment]
@hmoniz I have updated the code based on the most recent comments, please help take a look, thx!

**Review by hmoniz — APPROVED** [review]
LGTM. Thanks!

**Review by adlrocha — APPROVED** [review]
Thank you @cryptoAtwill , @hmoniz for pushing this to the finish-line. I think we are ready to merge, right? @cryptoAtwill would you mind fixing the conflicts before merging? Thanks!

**Comment by cryptoAtwill** [comment]
@adlrocha resolved and merging!

**Pull request merged**
